async-nats/src/connection.rs

```diff
@@ -40,254 +41,252 @@ pub(crate) struct Connection {
+            return Ok(Some(ServerOp::Error(ServerError::new(message))));
+        }
 
-                self.buffer.advance(len + 2);
+        if self.buffer.starts_with(b"INFO") {
+            let info = serde_json::from_slice(&self.buffer[5..len])
+                .map_err(|err| io::Error::new(io::ErrorKind::InvalidInput, err))?;
 
-                return Ok(Some(ServerOp::Info(Box::new(server_info))));
-            }
+            self.buffer.advance(len + 2);
 
-            return Ok(None);
+            return Ok(Some(ServerOp::Info(Box::new(info))));
         }
 
-        if self.buffer.starts_with(b"MSG ") {
-            if let Some(len) = self.buffer.find(b"\r\n") {
-                let line = std::str::from_utf8(&self.buffer[4..len]).unwrap();
-                let args = line.split(' ').filter(|s| !s.is_empty());
-                // TODO(caspervonb) we can drop this alloc
-                let args = args.collect::<Vec<_>>();
-
-                // Parse the operation syntax: MSG <subject> <sid> [reply-to] <#bytes>
-                let (subject, sid, reply_to, payload_len) = match args[..] {
-                    [subject, sid, payload_len] => (subject, sid, None, payload_len),
-                    [subject, sid, reply_to, payload_len] => {
-                        (subject, sid, Some(reply_to), payload_len)
-                    }
-                    _ => {
-                        return Err(io::Error::new(
-                            io::ErrorKind::InvalidInput,
-                            "invalid number of arguments after MSG",
-                        ));
-                    }
-                };
-
-                let sid = u64::from_str(sid)
-                    .map_err(|err| io::Error::new(io::ErrorKind::InvalidInput, err))?;
-
-                // Parse the number of payload bytes.
-                let payload_len = usize::from_str(payload_len)
-                    .map_err(|err| io::Error::new(io::ErrorKind::InvalidInput, err))?;
-
-                // Only advance if there is enough data for the entire operation and payload remaining.
-                if len + payload_len + 4 <= self.buffer.remaining() {
-                    let subject = subject.to_owned();
-                    let reply_to = reply_to.map(String::from);
-
-                    self.buffer.advance(len + 2);
-                    let payload = self.buffer.split_to(payload_len).freeze();
-                    self.buffer.advance(2);
-
-                    return Ok(Some(ServerOp::Message {
-                        sid,
-                        reply: reply_to,
-                        headers: None,
-                        subject,
-                        payload,
-                        status: None,
-                        description: None,
-                    }));
+        if self.buffer.starts_with(b"MSG") {
+            let line = str::from_utf8(&self.buffer[4..len]).unwrap();
```

Comment:
I think you want to include space inside quotes `"MSG "` otherwise the split below will create an extra "" (empty) arg, and `buffer[4..len]` eats an arbitrary character after the G

Reply:
Again, spot on 🙂